File: elixir/rational-numbers/lib/rational_numbers.ex
defmodule RationalNumbers do
  @type rational :: {integer, integer}

  @doc """
  Add two rational numbers
  """
  @spec add(x :: rational, y :: rational) :: rational
  def add({xa, xb}, {ya, yb}) do
    {xa * yb + xb * ya, xb * yb} |> reduce
  end

  @doc """
  Subtract two rational numbers
  """
  @spec subtract(x :: rational, y :: rational) :: rational
  def subtract({xa, xb}, {ya, yb}) do
    {xa * yb - xb * ya, xb * yb} |> reduce
  end

  @doc """
  Multiply two rational numbers
  """
  @spec multiply(x :: rational, y :: rational) :: rational
  def multiply({xa, xb}, {ya, yb}) do
    {xa * ya, xb * yb} |> reduce
  end

  @doc """
  Divide two rational numbers
  """
  @spec divide_by(num :: rational, den :: rational) :: rational
  def divide_by({num_a, num_b}, {den_a, den_b}) do
    {num_a * den_b, num_b * den_a} |> reduce
  end

  @doc """
  Absolute value of a rational number
  """
  @spec abs(x :: rational) :: rational
  def abs({a, b}) do
    {Kernel.abs(a), Kernel.abs(b)} |> reduce
  end

  @doc """
  Exponentiation of a rational number by an integer
  """
  @spec pow_rational(x :: rational, n :: integer) :: rational
  def pow_rational({a, b}, n) when n < 0 do 
    pow_rational({b, a}, Kernel.abs(n))
  end
  def pow_rational({a, b}, n) do
    {a ** n, b ** n} |> reduce
  end

  @doc """
  Exponentiation of a real number by a rational number
  """
  @spec pow_real(x :: integer, n :: rational) :: float
  def pow_real(x, n) do
    {a, b} = n
    x ** (a / b)
  end

  @doc """
  Reduce a rational number to its lowest terms
  """
  @spec reduce(x :: rational) :: rational
  def reduce({a, b}) when b < 0 do
    reduce({-a, -b})
  end
  def reduce({a, b}) do
      gcd = Integer.gcd(a, b)
      {div(a, gcd), div(b, gcd)}
  end
end

File: c/rational-numbers/rational_numbers.c
#include "rational_numbers.h"
#include <math.h>
#include <stdlib.h>

int gcd(int a, int b);

int gcd(int a, int b) {
  while (b != 0) {
    a %= b;
    a ^= b;
    b ^= a;
    a ^= b;
  }
  return a;
}

rational_t add(const rational_t r1, const rational_t r2) {
  return reduce((rational_t){
      .numerator =
          r1.numerator * r2.denominator + r2.numerator * r1.denominator,
      .denominator = r1.denominator * r2.denominator,
  });
}

rational_t subtract(const rational_t r1, const rational_t r2) {
  return reduce((rational_t){
      .numerator =
          r1.numerator * r2.denominator - r2.numerator * r1.denominator,
      .denominator = r1.denominator * r2.denominator,
  });
}

rational_t multiply(const rational_t r1, const rational_t r2) {
  return reduce((rational_t){
      .numerator = r1.numerator * r2.numerator,
      .denominator = r1.denominator * r2.denominator,
  });
}

rational_t divide(const rational_t r1, const rational_t r2) {
  return reduce((rational_t){
      .numerator = r1.numerator * r2.denominator,
      .denominator = r2.numerator * r1.denominator,
  });
}

rational_t absolute(const rational_t r) {
  return reduce((rational_t){
      .numerator = abs(r.numerator),
      .denominator = abs(r.denominator),
  });
}

rational_t exp_rational(const rational_t r, const int n) {
  rational_t r1;
  if (n > 0) {
    r1.numerator = pow(r.numerator, n);
    r1.denominator = pow(r.denominator, n);
  } else {
    r1.numerator = pow(r.denominator, -n);
    r1.denominator = pow(r.numerator, -n);
  }
  return reduce(r1);
}

float exp_real(const int n, const rational_t r) {
  return pow(pow(n, r.numerator), 1 / (double)r.denominator);
}

rational_t reduce(const rational_t r) {
  int divisor = gcd(r.numerator, r.denominator);
  int n = r.numerator / divisor;
  int d = r.denominator / divisor;
  if (d < 0) {
    n = -n;
    d = -d;
  }
  return (rational_t){
      .numerator = n,
      .denominator = d,
  };
}
